packages/7-todo-tdd/src/components/todo-app/TodoApp.test.tsx

```diff
@@ -0,0 +1,157 @@
+import { ReactWrapper } from 'enzyme';
+import { wrapperGenerator } from '../../styles/utils';
+import { TodoAppDriver } from './TodoApp.test.driver';
+import { s } from './TodoApp';
+import { Chance } from 'chance';
+import { TodoItemDriver } from '../todo-item/TodoItem.test.driver';
+import { todosService } from '../../services/TodoService';
+import { ITodoDTO } from 'fed-todo-journey_todo-common';
+
+const chance = new Chance();
+const h = wrapperGenerator('.');
+
+
+describe('app main components should be rendered', () => {
+    let mountedTodoApp: ReactWrapper;
+
+    beforeEach(async () => {
+        const todosDriver: TodoAppDriver = new TodoAppDriver();
+        await todosDriver.waitForAppToUpdate();
+        mountedTodoApp = todosDriver.getAppComponent();
+    });
+
+    it('app should be rendered', () => {
+        expect(mountedTodoApp.exists()).toBe(true);
+    });
+
+    it('todo input should be rendered', () => {
+        expect(mountedTodoApp.find(h(s.todo__inputContainer__input)).exists()).toBe(true);
+    });
+
+    it('todo list should be rendered', () => {
+        expect(mountedTodoApp.find(h(s.todo__list)).exists()).toBe(true);
+    });
+
+    it('add btn should be rendered', () => {
+        expect(mountedTodoApp.find(h(s.todo__inputContainer__addBtn)).exists()).toBe(true);
+    });
+});
+
+describe('checks todo list - add functionality', () => {
+    let appTestDriver: TodoAppDriver;
+
+    beforeEach(() => {
+        appTestDriver = new TodoAppDriver();
+    });
+
+    afterEach(async ()=>{
+        appTestDriver.removeTodosFromFakeTodosService();
+    })
+
+    it('should not add todo item with empty content', async () => {
+        appTestDriver.todoInputInsertContent('');
+        appTestDriver.clickOnAddBtn();
+        await appTestDriver.waitForAppToUpdate();
+        expect(appTestDriver.getTodosCount()).toBe(0);
+    });
+
+    it('should add todo item with content when pressing enter', async () => {
+        appTestDriver.todoInputInsertContent('example');
+        appTestDriver.pressEnterOnTodoInput();
+        await appTestDriver.waitForAppToUpdate()
+        expect(appTestDriver.getTodosCount()).toBe(1);
+    })
+
+    it('should add todo item with content when clicking on add btn', async () => {
+        appTestDriver.todoInputInsertContent('example');
+        appTestDriver.clickOnAddBtn();
+        await appTestDriver.waitForAppToUpdate()
+        expect(appTestDriver.getTodosCount()).toBe(1);
+    })
+
+    it('todo input should be empty after adding todo', async ()=>{
+        appTestDriver.todoInputInsertContent('example');
+        appTestDriver.clickOnAddBtn();
+        await appTestDriver.waitForAppToUpdate()
+        expect(appTestDriver.getTodoInputContent()).toBe(''); 
+    })
+})
+
+describe('app should render fetched todos properly', () => {
+    let content: string;
+    let appTestDriver: TodoAppDriver;
+
+    beforeAll(async () => {
+        content = chance.word();
```

Comment:
in general i really dislike `before` functions. they're good for setting up static things like `new Driver()`. But they're not good for setting up a common state. it makes the other tests unreadable.

what should u do instead?
write small functions at the bottom of the describe for these setups. 
for example 
`givenTodos([chance.word(),chance.word() ])`
`givenTodo(chance.word())`

then it's not a problem to call it from within the test itself and it makes the test so much more readable